### ISSUE TITLE
multi-bucket deterministic wallets (beyond receiving/change)

### ISSUE THREAD

**Issue by markblundeberg** [Electron-Cash/Electron-Cash#1256]
Currently Electron Cash internals support two address buckets -- receiving, and change. Discussion around SLP tokens has made it apparent that it would be nice if more buckets (derivation paths) could be supported -- SLP receiving, SLP change. That would allow SLP-aware wallets to store the primary BCH balance on a standard derivation path (that can be shared with SLP-unaware wallets) and keep the SLP tokens (with minimal BCH) safely hidden away on a separate path, where unaware wallets will never look. Only simpleledger: addresses would be shown for SLP-bucket addresses, and only bitcoincash: addresses would be shown for the BCH-bucket addresses.

The primary work here would involve digging into the guts of the backend code and UI code to just allow for this possibility, which is fairly involved since so far everything seems to be hardcoded for one bucket or two buckets.

Note:
* During the refactor process, it should be trivial to also allow each bucket to have a distinct xpub -- this gives the freedom of putting SLP onto a separate hardened path (44'/245'/0'/0/k), while still allowing to put SLP on an unhardened relative change index path (44'/145'/0'/5/k for example).

**Comment by markblundeberg** [comment]
@imaginaryusername points out this is may also be helpful for shuffles (hide shuffled coins from regular wallets so they don't mix with regular coins...?)

**Comment by cculianu** [comment]
> During the refactor process, it should be trivial to also allow each bucket to have a distinct xpub -- this gives the freedom of putting SLP onto a separate hardened path (44'/245'/0'/0/k), while still allowing to put SLP on an unhardened relative change index path (44'/145'/0'/5/k for example).

Yeah I think this may fit nicely (or not? have to check) with the existing wallet model of 1 wallet having multiple possible "keystores".

But yes -- most of the UI and wallet logic will need to be looked at and possibly refactored/modified/etc.

It's not trivial but not rocket science.  Definitely doable.


Thanks for opening the issue so it's on the radar.

**Comment by imaginaryusername** [comment]
Updating per recent discussion: It seems like an easier way to do this might be to have separate "wallet" files, perhaps with different suffixes, and load them together but as separate wallets (i.e. there can be wallet1 and wallet1.slp). They can then interact strictly through coinchooser, the GUI and change address fetching. 

Don't know which path is more effort and/or tech debt, but figured we should keep track of the evolving context.
